cpp/src/groupby/sort/group_single_pass_reduction_util.cuh

```diff
@@ -60,70 +63,72 @@ struct reduce_functor {
+    } else {
+      result = make_fixed_width_column(data_type(type_to_id<ResultType>()), 
+                                       num_groups,
+                                       mask_state::UNALLOCATED,
+                                       stream, mr);
+    }
 
     if (values.size() == 0) {
       return result;
     }
 
-    if (values.nullable()) {
-      T default_value = OpType::template identity<T>();
-      auto device_values = column_device_view::create(values);
-      auto val_it = cudf::experimental::detail::make_null_replacement_iterator(
-                        *device_values, default_value);
-
-      // Without this transform, thrust throws a runtime error
-      auto it = thrust::make_transform_iterator(val_it,
-                          [] __device__ (auto i) -> ResultType { return i; });
-      
-      thrust::reduce_by_key(rmm::exec_policy(stream)->on(stream),
-        // Input keys
-          group_labels.begin(), group_labels.end(),
-        // Input values
-          it,
-        // Output keys
-          thrust::make_discard_iterator(),
-        // Output values
-          result->mutable_view().begin<ResultType>(),
-        // comparator and operation
-          thrust::equal_to<size_type>(), op);
-
-      auto result_view = mutable_column_device_view::create(*result);
-      auto group_size_view = column_device_view::create(group_sizes);
-
-      thrust::for_each_n(rmm::exec_policy(stream)->on(stream),
-        thrust::make_counting_iterator(0), group_sizes.size(),
-        [d_result=*result_view, d_group_sizes=*group_size_view]
-        __device__ (size_type i){
-          size_type group_size = d_group_sizes.element<size_type>(i);
-          if (group_size == 0)
-            d_result.set_null(i);
-          else
-            d_result.set_valid(i);
-        });
-    } else {
-      auto it = thrust::make_transform_iterator(values.data<T>(),
-                          [] __device__ (auto i) -> ResultType { return i; });
-      thrust::reduce_by_key(rmm::exec_policy(stream)->on(stream),
-                            group_labels.begin(), group_labels.end(),
-                            it, thrust::make_discard_iterator(),
-                            result->mutable_view().begin<ResultType>(),
-                            thrust::equal_to<size_type>(), op);
-    }
+    thrust::fill(rmm::exec_policy(stream)->on(stream),
+      result->mutable_view().begin<ResultType>(), 
+      result->mutable_view().end<ResultType>(),
+      OpType::template identity<ResultType>());
+
+    auto resultview = mutable_column_device_view::create(result->mutable_view());
+    auto valuesview = column_device_view::create(values);
+
+    thrust::for_each_n(rmm::exec_policy(stream)->on(stream),
+      thrust::make_counting_iterator(0), values.size(),
+      [
+        d_values = *valuesview,
+        d_result = *resultview,
+        dest_indices = group_labels.data().get()
+      ] __device__ (auto i) {
+        if (d_values.is_valid(i))
```

Comment:
I suspect this is going to be a lot slower than a reduce by key due to atomic contention.

Reply:
It might be. But I don't see another option because `reduce_by_key` is erroneous. See this thread https://github.com/rapidsai/cudf/pull/3597#discussion_r363983804. Also, benchmarks reveal that this agg is a minuscule cost in the whole operation.